cpp/ql/src/experimental/Security/CWE/CWE-193/InvalidPointerDeref.ql

```diff
@@ -495,65 +480,89 @@ class PathSinkNode extends MergedPathNode, TPathNodeSink {
   }
 }
 
+/**
+ * A configuration that represents the full dataflow path all the way from
+ * the allocation to the dereference. We need this final dataflow traversal
+ * to ensure that the transition from the sink in `AllocToInvalidPointerConfig`
+ * to the source in `InvalidPointerToDerefFlow` did not make us construct an
+ * infeasible path (which can happen since the transition from one configuration
+ * to the next does not preserve information about call contexts).
+ */
+module FinalConfig implements DataFlow::StateConfigSig {
+  newtype FlowState =
+    additional TInitial() or
+    additional TPointerArith(PointerArithmeticInstruction pai, int delta) {
+      invalidPointerToDerefSource(_, pai, _, delta)
+    }
+
+  predicate isSource(DataFlow::Node source, FlowState state) {
+    state = TInitial() and
+    exists(DataFlow::Node derefSource |
+      invalidPointerToDerefSource(source, _, derefSource, _) and
+      InvalidPointerToDerefFlow::flow(derefSource, _)
+    )
+  }
+
+  predicate isSink(DataFlow::Node sink, FlowState state) {
+    exists(DataFlow::Node source, PointerArithmeticInstruction pai, int delta |
+      InvalidPointerToDerefFlow::flow(source, sink) and
+      invalidPointerToDerefSource(_, pai, source, delta) and
+      state = TPointerArith(pai, delta)
+    )
+  }
+
+  predicate isBarrier(DataFlow::Node n, FlowState state) { none() }
+
+  predicate isAdditionalFlowStep(
```

Comment:
Is this actually doing anything since we're using `invalidPointerToDerefSource` as a source anyway?

Reply:
We're starting the flow at the allocation just like before. Note that the first column of `invalidPointerToDerefSource` (which is the one we use to mark the source in `FinalConfig`):
```ql
predicate isSource(DataFlow::Node source, FlowState state) {
  state = TInitial() and
  exists(DataFlow::Node derefSource |
    invalidPointerToDerefSource(source, _, derefSource, _) and
    InvalidPointerToDerefFlow::flow(derefSource, _)
  )
}
```
is the dataflow node that represents the allocation:
```ql
predicate invalidPointerToDerefSource(
  DataFlow::Node source1, PointerArithmeticInstruction pai, DataFlow::Node derefSource, int delta
) {
  exists(
    AllocToInvalidPointerFlow::PathNode1 pSource1, AllocToInvalidPointerFlow::PathNode1 pSink1,
    AllocToInvalidPointerFlow::PathNode2 pSink2, DataFlow::Node sink1, DataFlow::Node sink2,
    int delta0
  |
    pragma[only_bind_out](pSource1.getNode()) = source1 and
    // ...
    AllocToInvalidPointerFlow::flowPath(pSource1, _, pragma[only_bind_into](pSink1),
      pragma[only_bind_into](pSink2)) and
    // ...
  )
}
```

Reply:
I see now. Why is all this complexity needed, and why can we just use the source from `AllocToInvalidPointerConfig`?

Reply:
It needs to be a source that actually flows to the sink. If we just write `AllocToInvalidPointerConfig::isSourcePair(source1, _, _, _, _)` we'll mark all allocations as sources irregardless of all the filtering that the rest of the query has done.

Reply:
But it seems that should be handled in the additional flow step and now here.

Reply:
We're currently starting to end up with several ways of doing the same thing it seems, see also `joinOn1`.

Reply:
But there's no reason to include the sources if we know they'll never reach the sink, though, right?

Reply:
> We're currently starting to end up with several ways of doing the same thing it seems, see also `joinOn1`.

`joinOn1` has been deleted, though. But you're saying that (when it existed) it did the same thing in a different way?

Reply:
Having a query that is actually readable and not very opaque with regard to what the sources are in various configurations sound like a very good reason to me. The query currently fails utterly in this respect.

Reply:
I'm in the process of writing detailed documentation for the query right now. It sounds like I should delay this PR until this documentation is in place.

Reply:
This is not a documentation issue. This is a query structuring issue.

Reply:
> with more and more functionality piling into `invalidPointerToDerefSource`.

This PR doesn't do anything to `invalidPointerToDerefSource` other than renaming a variable and moving another variable from `exists` to a parameter.

> This is a query structuring issue.

In that case, I'll think about how we can change the structure of the query, then.

Reply:
> This PR doesn't do anything to invalidPointerToDerefSource other than renaming a variable and moving another variable from exists to a parameter.

Allowing the predicate to be used in one more non-obvious way. Maybe we should start by actually properly naming the predicate (because it being a source of one of the dataflow configurations is probably the least relevant of all its uses at the moment), and also giving its parameters more descriptive names.

More general, every name that ends in a number should probably get a more descriptive name that does not involve number.

Reply:
Those are definitely some easy changes to make. I've created an internal issue for these things, and I suggest we delay this PR (as well as https://github.com/github/codeql/pull/13725) until this has been completed.

Reply:
I wonder is we could actually properly simplify the query based on the work you did here. In particular:
* Could we get rid of `TMergedPathNode` by adding another state to step to `FinalConfig` that goes to the final dereference?
* Could we get rid of `InvalidPointerToDerefConfig` by merging it's functionality into `FinalConfig`?

Reply:
I was able to get rid of `TMergedPathNode` in 5e06043120c8a2f609112709f397d3e203a717a6. I'm still trying to think about how we can get rid of `InvalidPointerToDerefConfig` by merging its functionality into `FinalConfig`.